### ISSUE TITLE
Support for ExAws credentials in SES adapter?

### ISSUE THREAD

**Issue by ascandella** [swoosh/swoosh#683]
This is not so much a specific feature request, but wanted to leave my thoughts here after spending afternoon getting Swoosh set up with SES for my AWS application.

The current method for configuring SES credentials (e.g. #361) is to pass them in as elixir application configuration (`:access_key` / `:secret`) on the adapter, which would generally happen in `runtime.exs` in a Phoenix application.

This works fine if you use static credentials, but this is not the optimal way to configure AWS permissions for services running inside AWS. Instead of generating (and securely storing / fetching, and maybe possibly even rotating) a single pair of keys that are valid until revoked, you can use the "instance role" (IAM role) of a running service (lambda, EC2 container, ECS service, etc.). Then, when that application needs to talk to another AWS service, it contacts a metadata service running on a special `169.254.169.254` server. This is generally abstracted away by every AWS library (including `ex_aws`).

I understand the reason for not taking on an `ex_aws` dependency for this library, but I imagine at least one other person may run into this limitation (where they don't want to provision a static keypair, instead delegating to the usual IAM strategy).

For my application, I've created a shim adapter that wraps the `AmazonSES` adapter provided by this library:

```elixir
defmodule ExAwsSESAdapter do                                                                                  
  @moduledoc """                                                 
  In order to support instance roles (rather than                                                       
  static keys passed through via environment variables), call `ExAws.Config.new`                                                    
  to respect the settings we've already set up for the app and pass those                                                           
  temporary credentials along to the `Swoosh.Adapters.AmazonSES`.                                                                   
  """                                                                                                                               
  use Swoosh.Adapter,                                                                                                               
    required_deps: [gen_stmp: :mimemail],                                                                                           
    required_config: [:region]                                                                                                      
                                                                                                                                    
  alias Swoosh.Email                                                                                                                
  alias Swoosh.Adapters.AmazonSES                                                                                                   
                                                                                                                                    
  @impl true                                                                                                                        
  def deliver(%Email{} = email, config \\ []) do                                                                                    
    credentials = ExAws.Config.new(:ses)                                                                                            
                                                                                                                                    
    config = add_credentials(config, credentials)                                                                                   
                                                                                                                                    
    email                                                                                                                           
    |> add_security_token(credentials)                                                                                              
    |> AmazonSES.deliver(config)                                                                                                    
  end                                                                                                                               
                                                                                                                                    
  defp add_credentials(config, credentials) do                                                                                      
    Keyword.merge(config,                                                                                                           
      access_key: credentials.access_key_id,                                                                                        
      secret: credentials.secret_access_key                                                                                         
    )                                                                                                                               
  end                                                                                                                               
                                                                                                                                    
  defp add_security_token(email, %{security_token: token}) do                                                                       
    Email.put_provider_option(email, :security_token, token)                                                                        
  end                                                                                                                               
                                                                                                                                    
  defp add_security_token(email, _), do: email                                                                                      
end
```

This puts a (short-lived) access key / secret key, as well as the `security_token` (if running from an instance profile), into the correct place that the `Swoosh.Adapters.AmazonSES` expects, without any static credentials existing.

Again, I understand not wanting to take on an additional dependency, so I'm mostly opening this issue (which you can feel free to close) so that another person may find it and have some time saved.

**Comment by princemaple** [comment]
Hi @ascandella , I'm actually fairly happy to accept this adapter. We have many [optional dependencies](https://github.com/swoosh/swoosh/blob/main/mix.exs#L68-L73) anyway.

We can require it on the adapter like: `required_deps: [gen_stmp: :mimemail, ex_aws: ExAws.Config]`

**Comment by ascandella** [comment]
@princemaple cool, do you think it makes sense to include in the existing `AmazonSES` adapter (and do something like `if Code.ensure_loaded?(ExAws) do ...` or have two separate adapters?

**Comment by princemaple** [comment]
I'm happy with it being standalone. That way we avoid clunky logics.

**Comment by princemaple** [comment]
Hi @ascandella, are you still planning to add this? I think with some links to the AWS docs added in the module docs, we are pretty much good to go.

**Comment by princemaple** [comment]
Published in v1.6.5 🥳 Thanks heaps @ascandella

**Comment by ascandella** [comment]
Sweet, thanks for the helpful replies and reviewing the PR so quickly! Just updated my app to 1.6.5 :)
